win_shortcut: Rewrite using AnsibleModule

dagwieers opened the pull request:
##### SUMMARY
Simple rewrite using AnsibleModule.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_shortcut

##### ANSIBLE VERSION
v2.8

dagwieers committed:
win_shortcut: Rewrite using AnsibleModule

jborean93 commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -5,33 +5,53 @@
+$src = $module.Params.src
+$dest = $module.Params.dest
+$state = $module.Params.state
+$args = $module.Params.args
```
This is dangerous, `$args` is a global variable set by PowerShell to contain the input arguments to the script/function.

dagwieers replied:
Right, we did it before as well. I guess because we don't rely on the original `$args` before.
I'll fix it.

dagwieers committed:
Avoid using $args, fix issue

jborean93 commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -5,33 +5,49 @@
+    $description = [System.Environment]::ExpandEnvironmentVariables($description)
+}
+
+$module.Result.changed = $false
```
Just an FYI, changed is automatically set on the return result to `$false`.

dagwieers replied:
Makes sense. Fixed.

jborean93 commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -47,13 +63,13 @@ If ($state -eq "absent") {
         } Catch {
             # Report removal failure
-            Fail-Json -obj $result -message "Failed to remove shortcut '$dest'. ($($_.Exception.Message))"
+            $module.FailJson("Failed to remove shortcut '$dest'. ($($_.Exception.Message))")
```
Change to `$module.FailJson("Failed to remove shortcut '$dest'. ($($_.Exception.Message))", $_)` so it will include the stacktrace and full error message when running with `-vvv`.

dagwieers replied:
Done.
Although named arguments, like `msg='....'` or `exception=$_` look nicer.

jborean93 replied:
It would but you can’t do that in PowerShell to .NET :(

jborean93 commented on lib/ansible/modules/windows/win_shortcut.ps1:
```diff
@@ -64,75 +80,75 @@ If ($state -eq "absent") {
             $ShortCut.Save()
         } Catch {
-            Fail-Json -obj $result -message "Failed to create shortcut '$dest'. ($($_.Exception.Message))"
+            $module.FailJson("Failed to create shortcut '$dest'. ($($_.Exception.Message))")
```
Add `$_` to the 2nd argument for FailJson to include a better error with `-vvv`.

dagwieers replied:
Done.

jborean93 reviewed (dismissed):
Looks good, just a quick change to some `FailJson()` calls. This method can take in an optional 2nd argument which is either an `ErrorRecord` (`$_`) or `Exception` (`$_.Exception`). When this is passed in, Ansible will include a proper error and stack trace in the traceback output to the user making it easier to debug.

dagwieers committed:
Small review fixes, and use 'arguments' parameter

dagwieers commented:
@jborean93 I implemented all your suggestions, and added one bonus.

IMO using `arguments'` is better than using `args` since we already have this parameter, so I renamed the parameter and added an alias.

jborean93 committed:
Update doc with new alias version change
